<a href="https://colab.research.google.com/github/jjyjung/ai/blob/gh-pages/tf_keras_CNN_MNIST.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import tensorflow as tf
from tensorflow.keras import layers, utils, Sequential
from tensorflow.keras.callbacks import EarlyStopping
from tensorflow.keras.datasets import mnist

In [2]:
# hyperparameters (실험변경 편의를 위하여 상단에 기재하였음)
batch_size = 100
epochs = 10
learning_rate = 0.1
dropout_rate = 0.1
patience = 10

In [3]:
# input image dimensions, output dimension
img_rows, img_cols = 28, 28
num_classes = 10

# the data, shuffled and split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
input_shape = (img_rows, img_cols, 1)

x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print('x_train shape:', x_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

# convert class vectors to binary class matrices
y_train = utils.to_categorical(y_train, num_classes)
y_test = utils.to_categorical(y_test, num_classes)

x_train shape: (60000, 28, 28, 1)
60000 train samples
10000 test samples


In [4]:
model = Sequential()
model.add(layers.Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=input_shape))
model.add(layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
model.add(layers.Dropout(dropout_rate))

model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
model.add(layers.Dropout(dropout_rate))
model.add(layers.Flatten())

model.add(layers.Dense(32, activation='relu'))
model.add(layers.Dropout(dropout_rate))
model.add(layers.Dense(num_classes, activation='softmax'))

model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 26, 26, 32)        320       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 13, 13, 32)       0         
 )                                                               
                                                                 
 dropout (Dropout)           (None, 13, 13, 32)        0         
                                                                 
 conv2d_1 (Conv2D)           (None, 11, 11, 64)        18496     
                                                                 
 max_pooling2d_1 (MaxPooling  (None, 5, 5, 64)         0         
 2D)                                                             
                                                                 
 dropout_1 (Dropout)         (None, 5, 5, 64)          0

In [5]:
# compile model
model.compile(
    loss = 'categorical_crossentropy',
    optimizer = 'Adam',
    metrics = ['accuracy'])

# train model
early_stop = EarlyStopping(monitor = 'val_loss', patience = 5, mode='min')
model.fit(x_train, y_train,
          batch_size = batch_size,
          epochs = epochs,
          validation_split = 0.2,
          callbacks = [early_stop], 
          shuffle = True,
          verbose = 1)

Epoch 1/10
480/480 [==============================] - 38s 77ms/step - loss: 0.3743 - accuracy: 0.8841 - val_loss: 0.0981 - val_accuracy: 0.9713
Epoch 2/10
480/480 [==============================] - 37s 77ms/step - loss: 0.1105 - accuracy: 0.9666 - val_loss: 0.0624 - val_accuracy: 0.9827
Epoch 3/10
480/480 [==============================] - 37s 77ms/step - loss: 0.0824 - accuracy: 0.9755 - val_loss: 0.0538 - val_accuracy: 0.9842
Epoch 4/10
480/480 [==============================] - 37s 77ms/step - loss: 0.0663 - accuracy: 0.9799 - val_loss: 0.0462 - val_accuracy: 0.9871
Epoch 5/10
480/480 [==============================] - 37s 77ms/step - loss: 0.0549 - accuracy: 0.9827 - val_loss: 0.0433 - val_accuracy: 0.9878
Epoch 6/10
480/480 [==============================] - 37s 77ms/step - loss: 0.0467 - accuracy: 0.9855 - val_loss: 0.0398 - val_accuracy: 0.9883
Epoch 7/10
480/480 [==============================] - 37s 76ms/step - loss: 0.0410 - accuracy: 0.9870 - val_loss: 0.0388 - val_accuracy:

"Epoch 3/10"의 의미는?

"38s 77ms/step"의 의미는?

"loss: 0.1105 - accuracy: 0.9666 - val_loss: 0.0624 - val_accuracy: 0.9827"의 의미는?

In [6]:
# evaluate model
score = model.evaluate(x_test, y_test, verbose = 1)
print('test_loss:', score[0])
print('test_acc:', score[1])

model.save("cnn_mnist.h5")

313/313 [==============================] - 3s 8ms/step - loss: 0.0302 - accuracy: 0.9898
test_loss: 0.03016565926373005
test_acc: 0.989799976348877
